File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

public class Solution {

    public static boolean checkPrefix(String[] phoneBook) {
        Map<Integer, List<String>> preDict = new HashMap<>();

        for (String num : phoneBook) {
            for (Integer key : preDict.keySet()) {
                for (String phone : preDict.get(key)) {
                    if (num.length() > key && num.substring(0, key).equals(phone)) {
                    return false;
                    }
                }
            }
        preDict.put(num.length(), new ArrayList<>(Arrays.asList(num)));
        }

    return true;
    }

    public static boolean solution(String[] phoneBook) {
        Arrays.sort(phoneBook);
        return checkPrefix(phoneBook);
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def check_prefix(phone_book):
    pre_dict = {}
    for num in phone_book:
        if not pre_dict:
            pre_dict[len(num)] = [num]
        else:
            for key in pre_dict.keys():
                for phone in pre_dict[key]:
                    if num[:key] == phone:
                        return False
            pre_dict[len(num)] = [num]
    return True

def solution(phone_book):
    phone_book.sort()
    answer = check_prefix(phone_book)
    
    return answer